### PR TITLE
feat: anchor service (part 2)

### PR DESCRIPTION
**Pull request by smrz2001**
This is the first PR for self-anchoring and includes:
- Anchor service machinery
- Merkle tree building
- Time Event building
- Traits for:
  - Anchor client (e.g. Event Service)
  - Transaction managers (e.g. remote CAS)

### CODE REVIEW

File: Cargo.toml
Comment (smrz2001) [inline Cargo.toml:221]: Please add more names here as suggestions, if interested.

File: anchor-service/src/time_event.rs
Comment (dav1do) [inline anchor-service/src/time_event.rs:34]: I wonder if it's worth adding an input struct here to make it slightly harder to mismatch all the cids.. it's still a bunch of cids but it was something like one of these two might be less error prone (if a bit tedious)

```rust
    struct BuildTimeEventInput<'a> {
        pub id: &'a Cid,
        pub prev: &'a Cid,
        pub root: &'a Cid,
        pub remote_path: &'a str,
        pub index: u64,
        pub count: u64,
    }

    struct BuildTimeEventInput2<'a> {
        pub id: Cid,
        pub prev: Cid,
        pub root: Cid,
        pub remote_path: &'a str,
        pub index: u64,
        pub count: u64,
    }

    fn build_it(input: BuildTimeEventInput) {
        todo!()
    }
    fn build_it2(input: &BuildTimeEventInput2) {
        todo!()
    }
    build_it(BuildTimeEventInput {
        id: &id,
        prev: &prev,
        root: &proof,
        remote_path: "",
        index,
        count,
    });
    // only need to assign input here if we still want access to the cids after
    let input = BuildTimeEventInput2 {
        id,
        prev,
        root: proof,
        remote_path: "",
        index,
        count,
    };

    build_it2(&input);
 ```

**Review by dav1do — COMMENTED** [review]
I didn't think too hard about the tree building stuff but poked through the tests. A few comments but no major concerns.

File: anchor-service/Cargo.toml
Comment (nathanielc) [inline anchor-service/Cargo.toml:1]: Can you add the `/anchor-service/` dir to the CODEOWNERS file so we establish reviewers for the rest of this code.

File: event/src/lib.rs
Comment (nathanielc) [inline event/src/lib.rs:6]: Why do anchor types live in the `ceramic-event` crate? I would have expected them to live in the `ceramic-anchor-service` crate.

File: event/src/unvalidated/event.rs
Comment (nathanielc) [inline event/src/unvalidated/event.rs:305]: These should not be public. They are intentionally private so that they cannot be used to accidentally construct nonsense events.

File: anchor-service/src/time_event.rs
Comment (nathanielc) [inline anchor-service/src/time_event.rs:43]: Why do we not use the builder API here?

File: event/src/anchor.rs
Comment (nathanielc) [inline event/src/anchor.rs:21]: Curious on the choice of an IndexMap? Should we be using an IndexMap anytime we have CID keys?

File: anchor-service/src/anchor_batch.rs
Comment (nathanielc) [inline anchor-service/src/anchor_batch.rs:22]: nit, can you add documentation to all the methods since this is public.

File: anchor-service/src/lib.rs
Comment (nathanielc) [inline anchor-service/src/lib.rs:1]: Can you add `#![warn(missing_docs)]` here so we document all public interfaces from this crate.

File: anchor-service/src/anchor_batch.rs
Comment (nathanielc) [inline anchor-service/src/anchor_batch.rs:19]: Who is expected to implement this trait? Something from the ceramic-service crate?

If so I think we should slightly tweak the semantics of the trait. Instead of AnchorClient we should call this FeedClient and instead of `get_anchor_requests` we should have `get_events(FeedOpts)` where FeedOpts lets you decide which kind of events to get from the Feed. This way the `ceramic-service` does not have to know or understand anything about anchors. Rather it remains specific to Ceramic events.  

This likely also means that the AnchorRequest type can remain private to the anchor-service crate.


Thoughts?

**Review by nathanielc — CHANGES_REQUESTED** [review]
Looking good, have some questions below.

File: anchor-service/src/anchor_batch.rs
Comment (AaronGoldman) [reply anchor-service/src/anchor_batch.rs:46]: yes we want the  first attempt right away

File: event/src/anchor.rs
Comment (AaronGoldman) [reply event/src/anchor.rs:21]: This is to keep the merkel tree nodes in order.  A hash map will randomize the order.
Non-deterministic order makes testing hard.
Sorted order you will need to use later tree nodes to use earlier ones. Not nesaraly a problem just pulling more nodes into memory as we validate the time events.

**Comment by stbrody** [comment]
Holding off on reviewing this until the first round of comments from Nathaniel and David are addressed, but would love to review this before it goes in

File: event/src/anchor.rs
Comment (nathanielc) [reply event/src/anchor.rs:21]: Ok so why not a simple BTreeMap in that case?

Comment (nathanielc) [reply event/src/anchor.rs:21]: Actually I think I misunderstood _order_. We need to keep the nodes in insertion order not just any stable order?

File: anchor-service/src/merkle_tree.rs
Comment (stbrody) [inline anchor-service/src/merkle_tree.rs:17]: what are "peaks" here?  TBH I feel like this whole function could use a lot more comments explaining what it's doing

File: event/src/anchor.rs
Comment (stbrody) [inline event/src/anchor.rs:16]: what about a node that has only has a child in one direction?  Don't we really need each node to be a pair of Option<Cid> so we can also use None when one branch is empty?

File: event/src/anchor.rs
Comment (stbrody) [inline event/src/anchor.rs:21]: comment on what the relationship is between the key and the value please

File: anchor-service/src/merkle_tree.rs
Comment (stbrody) [inline anchor-service/src/merkle_tree.rs:34]: I find this code confusing to understand.

Is there a reason we're using some fancy algorithm here instead of just the simple recursive balanced binary tree building algorithm taught in every college data structures class?  We know all the events up front, we don't need a streaming algorithm, and I think there's value in keeping this really simple and obviously correct for now.  If we want to move to a streaming approach to tree building we can do so in the future, but doing that would require changes to a lot more than just this function, so as long as we're building the whole tree in one go I'd strongly vote to keep the algorithm dead simple for now.

**Review by stbrody — CHANGES_REQUESTED** [review]
did a quick skim over this for some initial thoughts.  A big overarching request is more comments throughout.

File: anchor-service/src/anchor_batch.rs
Comment (AaronGoldman) [reply anchor-service/src/anchor_batch.rs:67]: Spawn a std::thread is higher up in the logic.  their will be an anchor loop that is once an hour calling the event service to get all unanchored events and do the anchoring. within that loop it is fine for everything to be serial.

1. get the anchor requests
2. build the tree
3. send root to CAS / ETH
4. poll for transaciton completion
5. build the proof.
5.1. verify proof. not implemented 
6. build the time events.
7. store the time events.
8. sleep until next hour tick. 

One tikio thread going aroind and around in the circle.

File: event/src/unvalidated/event.rs
Comment (AaronGoldman) [reply event/src/unvalidated/event.rs:305]: RawTimeEvent new is just a pass through and there are getters for every field. The fields are not optional. 
 RawTimeEvent is just a data class I don't see why we would not just have all public fields and no functions on the data class. except for `impl Debug for RawTimeEvent`

File: event/src/anchor.rs
Comment (AaronGoldman) [reply event/src/anchor.rs:16]: for parsing yes for building the tree we don't have any not full nodes.

Comment (stbrody) [reply event/src/anchor.rs:16]: how is that possible?  What if you have a number of events that isn't a power of 2?  Won't you always have some unbalanced nodes?

Comment (AaronGoldman) [reply event/src/anchor.rs:16]: image
The depth may not be balanced but there are no nulls.
The peeks are the place values so
peek[0] has 1 anchor requested CID or a null.
peek[1] has 2 anchor requested CID or a null.
peek[2] has 4 anchor requested CID or a null.
peek[3] has 8 anchor requested CID or a null.
peek[n] has 2**n anchor requested CID or a null.

then we just role up the side merging the pow2 trees.
If we kept the nulls we would get balanced depth but nulls in the roll up nodes.
Since we don't bother with the nodes that have an empty branch we get no nulls but shorter paths where the half empty nodes would have been,

File: anchor-service/src/anchor_batch.rs
Comment (smrz2001) [reply anchor-service/src/anchor_batch.rs:46]: Looks like the first tick always completes immediately (from the `interval` docs):
```
    /// use tokio::time;
    ///
    /// use std::time::Duration;
    ///
    /// #[tokio::main]
    /// async fn main() {
    ///     let mut interval = time::interval(Duration::from_millis(10));
    ///
    ///     interval.tick().await;
    ///     // approximately 0ms have elapsed. The first tick completes immediately.
    ///     interval.tick().await;
    ///     interval.tick().await;
    ///
    ///     // approximately 20ms have elapsed.
    /// }
```

File: anchor-service/src/time_event.rs
Comment (smrz2001) [reply anchor-service/src/time_event.rs:34]: We went ahead and inlined the function.

File: core/src/block.rs
Comment (smrz2001) [reply core/src/block.rs:8]: Yes, we removed the file.

File: event/src/anchor.rs
Comment (smrz2001) [reply event/src/anchor.rs:21]: Yeah, the order here is only used for debugging and expect tests. It makes it easier for us to know where there are issues because the remote nodes and local nodes remain in their respective groupings.

File: anchor-service/src/anchor_batch.rs
Comment (smrz2001) [reply anchor-service/src/anchor_batch.rs:19]: This feels worth a design discussion.

File: event/src/anchor.rs
Comment (stbrody) [reply event/src/anchor.rs:16]: Ah this is still with the MMR approach, I was assuming we switched to the simpler balanced tree building approach

File: anchor-service/src/merkle_tree.rs
Comment (smrz2001) [reply anchor-service/src/merkle_tree.rs]: Yes, will add more comments soon!

Comment (smrz2001) [reply anchor-service/src/merkle_tree.rs]: Added comments from @AaronGoldman that we'd put in a later PR.

File: anchor-service/src/merkle_tree.rs
Comment (stbrody) [reply anchor-service/src/merkle_tree.rs:37]: Just spoke about this with @smrz2001 and he realized that because we need to replace older events with newer events on the same stream before we build the tree, that means the set of events that go into the tree isn't even strictly append-only.  Which means it's not even totally clear how we could *ever* do streaming tree building.  So it seems that using the MMR algorithm is not giving us any benefit now and is fairly likely to *never* give us any benefit.  So I fairly strongly think we should rip out this MMR approach and just put in the extremely simple balanced tree building algorithm instead.

Comment (AaronGoldman) [reply anchor-service/src/merkle_tree.rs:37]: You mean something like 
```rs
fn build_merkle_tree_r(anchor_requests: &[AnchorRequest], nodes: &mut MerkleNodes) -> Result<Cid> {
    if anchor_requests.len() == 0 {
        return Err(anyhow!("Empty anchor_requests"));
    }
    if anchor_requests.len() == 1 {
        return Ok(anchor_requests[0].prev);
    }
    let top_power_of_2 = 1 << (63 - anchor_requests.len().leading_zeros());
    let split_point;
    if anchor_requests.len() == top_power_of_2 {
        split_point = top_power_of_2 / 2;
    } else {
        split_point = top_power_of_2;
    }
    let left = build_merkle_tree_r(&anchor_requests[..split_point], nodes)?;
    let right = build_merkle_tree_r(&anchor_requests[split_point..], nodes)?;
    let node = vec![Some(left),Some(right)];
    let node_cid = node.to_cid()?;
    nodes.insert(node_cid, node);
    Ok(node_cid)
}
```
That should build the same tree.

File: anchor-service/src/time_event.rs
Comment (AaronGoldman) [reply anchor-service/src/time_event.rs]: We have a proof object CID and the builder API takes the fields. We would need to parse the proof block and plum through the fields from the CAS response just to eventually re assemble in the builder api and calculate the same CID. Assuming it round trips witch it should.

**Review by nathanielc — CHANGES_REQUESTED** [review]

File: anchor-service/src/anchor_batch.rs
Comment (AaronGoldman) [reply anchor-service/src/anchor_batch.rs:23]: The `impl Store for CeramicEventService` is in part 5. That is where I would like to make the decision on where we hydrate to time events and who is responsible for storing the high water mark. As part 5 is written today the high water mark is stored in the event service we will probably store that somewhere else.

`get_anchor_request` is renamed  `local_sourced_data_events` in part 3 but the real logical isolation will probably be in part 5 when the decision about where to store the high water mark drives the final trait definition.

https://github.com/ceramicnetwork/rust-ceramic/pull/488/files#diff-148a932282cead8137f562abc4a57762ed173ccebd8d74361766ec0df7948d22R424

`$1`  new_high_water_mark (i64)
```sql
INSERT INTO ceramic_one_anchor_high_water_mark (high_water_mark) VALUES ($1);
```

`$1` did:key for the local node
`$2` the limit
```sql
SELECT order_key, init_cid, cid, rowid
FROM ceramic_one_event
WHERE
  source = $1
  AND ROWID > (SELECT MAX(high_water_mark) FROM ceramic_one_anchor_high_water_mark)
ORDER BY rowid
LIMIT $2;
```

Comment (nathanielc) [reply anchor-service/src/anchor_batch.rs:23]: >The impl Store for CeramicEventService is in part 5. That is where I would like to make the decision on where we hydrate to time events and who is responsible for storing the high water mark.

Why does the ceramic-service what to store the high water mark for the anchor service? I would expect the anchor service to have its own place to store the highwater mark separate from the ceramic-service. Perhaps the name `anchor_service::Store` is incorrect here in that it implies its the interface for storing state on behalf of the anchor service. I see this trait as a way access to the event service. Perhaps we should call this trait `EventService`? 

And do we need another mechnism for storing data on behalf of the anchor service. For example a new sqlite connection and its own anchor store struct that can store and retrieve the high water mark?

Comment (nathanielc) [reply anchor-service/src/anchor_batch.rs:23]: To expand a bit on the architecture I am after...

I would like to have a set of services and each service manages its own state (i.e. database). The services can depend on each other via public APIs etc. In this PR we are adding a new service that depends on the event service. This means the anchor service needs it a way to manage its own state. It should not delegate management of its state to the event service. The anchor service has very little state that it needs to store (i.e. only the high water mark). However that still means that there should be some kind of struct private to the anchor service crate that manages storing and retrieving the high water mark. I expect this is a sqlite table, and it can be the same sqlite connection that the event service uses but the logic to store and retrieve that state lives inside the anchor service.

When we inevitably want to migrate how we store the high water mark, we will have to modify any code that access that storage directly. Its best if all of that code lives behind the anchor service API boundary. 

As the PR is currently structured its intended that logic lives inside the ceramic-service crate and I explicitly do not want that architecture.

Comment (AaronGoldman) [reply anchor-service/src/anchor_batch.rs:23]: * Part 2: build trees from anchor requests and build Time Events from proofs.
* Part 3: use the remote CAS to make proofs from root CIDs. All the auth and api stuff to talk to unmodified CAS.
* Part 4: all the changes to EventService so it can find the events that anchor service wants to anchor.  e.g. source and StreamID for events.
* Part 5: starting the loop, command line args, persisting Time events and High Water Mark

I think we can answer those questions in part 5 this is mostly here as a `todo!(get_anchor_requests, put_time_events)`
and to have something to for the MockStore so I could implement and test trees building and Time event building.

File: anchor-service/src/time_event.rs
Comment (stbrody) [inline anchor-service/src/time_event.rs:39]: I don't quite follow this.  I think the example text below the diagram is saying that the example is with `index` as 10 and `count` as 14?  Can we make that explicit in the comment?  Also, what does the "e.g. 5,6" mean here?

File: anchor-service/src/merkle_tree.rs
Comment (stbrody) [inline anchor-service/src/merkle_tree.rs:14]: is there any unit test coverage of this?  The MMR logic here is complex enough that I think some targeted unit tests would make me feel a lot better about it.

**Review by stbrody — APPROVED** [review]
main comment is to add a unit test for merkle tree building, otherwise lgtm

File: anchor-service/src/anchor_batch.rs
Comment (AaronGoldman) [reply anchor-service/src/anchor_batch.rs:17]: Both actually. The high water mark will be in a table owned by the anchor service and the get data events that originate from this node less then high water mark will be calling a function on  CeramicEventService.

File: anchor-service/src/merkle_tree.rs
Comment (smrz2001) [reply anchor-service/src/merkle_tree.rs:14]: We were in the middle of merging this PR when we saw this comment :(

Create a ticket under SA Ph1 for this:
https://linear.app/3boxlabs/issue/AES-343/add-more-exhaustive-merkle-tree-tests

**Pull request merged**
